File: solver-ms/solver.py
import os
import pika
import json
import subprocess
import tempfile

def solve_problem(ch, method, properties, body):
    try:
        problem_data = json.loads(body)
        input_file_content = problem_data['input_file']
        param1 = str(problem_data['param1'])
        param2 = str(problem_data['param2'])
        param3 = str(problem_data['param3'])

        # Create a temporary file to save the input JSON content
        with tempfile.NamedTemporaryFile(mode='w+', delete=False, suffix='.json') as temp_file:
            temp_file.write(input_file_content)
            temp_file_path = temp_file.name

        # Call the vrpSolver.py script
        result = subprocess.run(['python', 'vrpSolver.py', temp_file_path, param1, param2, param3], capture_output=True)
        print(" [x] Result: %s" % result.stdout)

        # Clean up the temporary file
        os.remove(temp_file_path)

        ch.basic_ack(delivery_tag=method.delivery_tag)
    except KeyError as e:
        print(f"KeyError: Missing key {e} in message {body}")
        ch.basic_ack(delivery_tag=method.delivery_tag)
    except Exception as e:
        print(f"Error processing message: {e}")
        ch.basic_ack(delivery_tag=method.delivery_tag)

credentials = pika.PlainCredentials('user', 'password')
connection = pika.BlockingConnection(pika.ConnectionParameters('rabbitmq', 5672, '/', credentials))
channel = connection.channel()

channel.queue_declare(queue='problem_queue')

channel.basic_consume(queue='problem_queue', on_message_callback=solve_problem)

print(' [*] Waiting for messages. To exit press CTRL+C')
channel.start_consuming()

File: solver-ms/solver.js
const amqp = require('amqplib');
require('dotenv').config();
const { exec } = require('child_process');
const fs = require('fs');
const os = require('os');
const path = require('path');

const rabbitmq = process.env.RABBITMQ_URL;
let channel = null;

let processing = false;
const messageQueue = [];

async function processMessage(channel) {
  if (messageQueue.length === 0) {
    processing = false;
    return;
  }

  processing = true;
  const msg = messageQueue.shift();

  if (msg !== null) {
    const content = msg.content.toString();
    console.log(" [x] Received %s", content);

    // Parse the JSON message
    const data = JSON.parse(content);
    const problemID = data._id;
    const inputFileContent = data.input_file;
    const param1 = data.param1;
    const param2 = data.param2;
    const param3 = data.param3;
    const createdBy = data.createdBy;

    // Create a temporary file for the input JSON
    const tempFilePath = path.join(os.tmpdir(), 'input_data.json');
    fs.writeFileSync(tempFilePath, inputFileContent);

    // Construct the command
    const command = `python vrpSolver.py ${tempFilePath} ${param1} ${param2} ${param3}`;

    // Update the problem status
    publishToQueue("status_queue", { id: problemID, newStatus: "in-progress" });
    console.log("Status updated to \"In-Progress\"");

    console.log("Executing command:", command);

    // Execute the command
    exec(command, (error, stdout, stderr) => {
      if (error) {
        console.error(`Error executing command: ${error.message}`);
      } else if (stderr) {
        console.error(`stderr: ${stderr}`);
      } else {
          try {
            output_file = stdout;
            publishToQueue("solvedProblems", { createdBy, output_file, problemID });
          } catch (error) {
              console.error('Error saving results:', error);
          } 
          console.log(`Mr/Mrs ${createdBy} here are your results: ${stdout}`);
      }

      // Acknowledge the message
      channel.ack(msg);

      // Process the next message in the queue
      processMessage(channel);
    });
  } else {
    processing = false;
  }
}

async function connectRabbitMQ() {
  for (let attempts = 1; attempts <= 5; attempts++) {
    try {
      const connection = await amqp.connect(rabbitmq);
      channel = await connection.createChannel();
      const queue = 'problem_queue';
      const queue1 = 'solvedProblems';
      const queue2 = 'status_queue';
  
      await channel.assertQueue(queue, { durable: true });
      await channel.assertQueue(queue1, { durable: true });
      await channel.assertQueue(queue2, { durable: true });
      console.log("Connected to RabbitMQ");
  
      channel.consume(queue, async (msg) => {
        if (msg !== null) {
          messageQueue.push(msg);
          if (!processing) {
            processMessage(channel);
          }
        }
      });
      return;
    } catch (error) {
      console.error(`Failed to connect to RabbitMQ (attempt ${attempts} of 5)`, error);
      if (attempts === 5) {
        console.error("Max retries reached. Exiting.");
        process.exit(1);
      } else {
        console.log(`Retrying in 5 seconds...`);
        await new Promise(res => setTimeout(res, 5000));
      }
  }
  }
}

async function publishToQueue (queueName, message) {
  if (!channel) {
    console.error('Channel not set. Call connectRabbitMQ() first.');
    return;
  }
  channel.assertQueue(queueName, { durable: true });
  channel.sendToQueue(queueName, Buffer.from(JSON.stringify(message)));
};

connectRabbitMQ();
